### PR TITLE
Adopt `TrustAnchorIdentifiers` from v03 of MTC draft

### PR DESCRIPTION
**Pull request by pohlm01**
This solves #3

### CODE REVIEW

File: ca/ca.go [hunk 1]
```diff
@@ -30,6 +30,7 @@ var (
 
 type NewOpts struct {
 	IssuerId   string
```
Comment (bwesterb) [inline ca/ca.go:32]: `IssuerId` needs to be removed.

File: mtc.go [hunk 2]
```diff
@@ -1590,6 +1564,23 @@ func (tai *TrustAnchorIdentifier) UnmarshalText(text []byte) error {
 		}
 	}
 	*tai = buf.Bytes()
```
Comment (bwesterb) [inline mtc.go:1566]: Needs length check.

File: mtc.go [hunk 3]
```diff
@@ -1527,12 +1482,43 @@ func NewMerkleTreeProof(batch *Batch, index uint64, path []byte) *MerkleTreeProo
...
-// A Trust Anchor Identifier (TAI) is used to identify a CA, or a specific batch.
+// A TrustAnchorIdentifier (TAI) is used to identify a CA, or a specific batch.
 //
 // TAI are OIDs relative to the Private Enterprise Numbers (PEN)
-// arc  1.3.6.1.4.1.
+// arc 1.3.6.1.4.1.
 type TrustAnchorIdentifier []byte
 
+func (tai *TrustAnchorIdentifier) ProofType() ProofType {
+	// TODO make this dependent on the OID
```
Comment (bwesterb) [inline mtc.go:1492]: We can't see it from the OID itself: we need some lookup. A caller will have a list of known CAs that include OID, public key, and proof type. The question is how to do this elegantly in the API. Perhaps we have a MTCContext struct, which when created can be passed with a `CAStore` interface that has a OID -> CAParams function.

File: mtc.go [hunk 4]
```diff
@@ -670,11 +623,13 @@ func (p *CAParams) newTreeHeads(prevHeads, root []byte) ([]byte, error) {
 	return append(prevHeads[HashLen:len(prevHeads)], root...), nil
 }
 
-func (batch *Batch) Anchor() *MerkleTreeTrustAnchor {
-	return &MerkleTreeTrustAnchor{
-		issuerId:    batch.CA.IssuerId,
-		batchNumber: batch.Number,
+func (batch *Batch) Anchor() *TrustAnchorIdentifier {
+	tai := TrustAnchorIdentifier{}
+	err := tai.UnmarshalText([]byte(fmt.Sprintf("%s.%d", batch.CA.IssuerOID, batch.Number)))
```
Comment (bwesterb) [inline mtc.go:628]: I guess it works. Perhaps neater to add a function on `TrustAnchorIdentifer` which does this more directly.

File: mtc.go [hunk 5]
```diff
@@ -515,8 +459,16 @@ func (p *CAParams) UnmarshalBinary(data []byte) error {
 		err           error
 	)
 
-	if !s.ReadUint8LengthPrefixed((*cryptobyte.String)(&issuerBuf)) ||
-		!s.ReadUint16((*uint16)(&sigScheme)) ||
+	if !s.ReadUint8LengthPrefixed((*cryptobyte.String)(&issuerBuf)) {
+		return ErrTruncated
+	}
+	tai := TrustAnchorIdentifier{}
+	err = tai.unmarshal(&s)
```
Comment (bwesterb) [inline mtc.go:466]: You don't check if `s` is empty.

File: mtc.go [hunk 6]
```diff
@@ -133,34 +134,19 @@ const (
...
-
 type Proof interface {
-	TrustAnchor() TrustAnchor
+	TrustAnchorIdentifier() *TrustAnchorIdentifier
 	Info() []byte
 }
 
 type MerkleTreeProof struct {
-	anchor *MerkleTreeTrustAnchor
+	anchor *TrustAnchorIdentifier
```
Comment (bwesterb) [inline mtc.go:143]: Why pointer?

File: mtc.go [hunk 7]
```diff
@@ -18,9 +18,10 @@ import (
 	"golang.org/x/crypto/cryptobyte"
 )
 
-// Public parameters of a Merkle Tree CA
+// CAParams holds the public parameters of a Merkle Tree CA
 type CAParams struct {
 	IssuerId           string
+	IssuerOID          TrustAnchorIdentifier
```
Comment (bwesterb) [inline mtc.go:24]: As IssuerId will be removed, we can just call this `Issuer`.

**Review by bwesterb — CHANGES_REQUESTED** [review]
Thanks for this!

**Comment by pohlm01** [comment]
Thanks a lot for this meaningful review! I'll add more commits and let you know when I think it might be in a mergeable state.

File: mtc.go (hunk 3 above)
Comment (pohlm01) [reply mtc.go]: I've built a simple construction, such that the CA params essentially are the only entry in the 'root store'. Let me know if you would like something more elaborate, or feel free to change it yourself.

File: mtc.go [hunk 8]
```diff
@@ -426,14 +425,7 @@ func (p *CAParams) ActiveBatches(dt time.Time) BatchRange {
...
 	var b cryptobyte.Builder
-	b.AddUint8LengthPrefixed(func(b *cryptobyte.Builder) {
-		b.AddBytes([]byte(p.IssuerId))
-	})
-	tai, err := p.IssuerOID.MarshalBinary()
-	if err != nil {
-		return nil, err
-	}
-	b.AddBytes(tai)
+	b.AddBytes(p.Issuer)
```
Comment (bwesterb) [inline mtc.go]: Shouldn't this be length prefixed?

File: README.md [hunk 9]
```diff
@@ -107,7 +107,7 @@ For every batch, the CA signs that root together with all the roots
 Let's create an MTC CA.
 
 ```
-$ mtc ca new --batch-duration 5m --lifetime 1h my-mtc-ca ca.example.com/path
+$ mtc ca new --batch-duration 5m --lifetime 1h 1230.12.15 ca.example.com/path
```
Comment (bwesterb) [inline README.md:110]: 1230 is `McAfee`.

Comment (bwesterb) [reply README.md:110]: You can use 62253 for the example.

File: ca/ca.go [hunk 10]
```diff
@@ -29,7 +29,7 @@ var (
 )
 
 type NewOpts struct {
-	IssuerId   string
+	Issuer     mtc.OID
```
Comment (bwesterb) [inline ca/ca.go:32]: This is not an OID, as OIDs are encoded differently. Instead it's an *arc under the PEN OID*, but that doesn't really roll off the tongue either.

File: mtc.go [hunk 11]
```diff
@@ -369,68 +321,55 @@ func (c *BikeshedCertificate) MarshalBinary() ([]byte, error) {
...
+		return nil, fmt.Errorf("failed to marshal Assertion: %w", err)
 	}
 	b.AddBytes(buf)
-	b.AddUint16(uint16(c.Proof.TrustAnchor().ProofType()))
-	b.AddUint8LengthPrefixed(func(b *cryptobyte.Builder) {
-		b.AddBytes(c.Proof.TrustAnchor().Info())
-	})
+
+	buf, err = c.Proof.TrustAnchorIdentifier().MarshalBinary()
+	b.AddBytes(buf)
```
Comment (bwesterb) [inline mtc.go:329]: Length prefix is missing

**Comment by bwesterb** [comment]
> Thanks a lot for this meaningful review! I'll add more commits and let you know when I think it might be in a mergeable state.

Good progress. There are still a couple of issues — see my comments, but they shouldn't be too hard to fix.

File: ca/ca.go (hunk 10 above)
Comment (bwesterb) [reply ca/ca.go:32]: Both 1234.1.2 and 1234.1 can be trust anchor identifiers in the more general context of [TLS Trust Anchor Identifiers](https://www.ietf.org/archive/id/draft-beck-tls-trust-anchor-ids-00.html) so we don't really need the two different types. If you keep it split, then I'd rename `OID` to `RelativeOID`, `PENArc` or something similar.

File: mtc.go (hunk 11 above)
Comment (pohlm01) [reply mtc.go:329]: The `MarshalBinary()` method of the `tai` does encode the OID with the TLS style length prefix. From how I understand the I-D, there is no other, outer length encoding needed. See [Section 5.5.3](https://datatracker.ietf.org/doc/html/draft-davidben-tls-merkle-tree-certs-03#section-5.5.3) and [`mct.go:1603`](https://github.com/pohlm01/mtc/blob/main/mtc.go#L1603) for reference.

Comment (bwesterb) [reply mtc.go:329]: The definition is

```
opaque TrustAnchorIdentifier<1..2^8-1>
```

As you can read from [section 4.3 from RFC 5246](https://datatracker.ietf.org/doc/html/rfc5246#section-4.3), the angular brackets indicate a length prefixed value.

Comment (pohlm01) [reply mtc.go:329]: Agreed. Maybe I'm just confusing something here, but in my mind, this is taken care of in [`mct.go:1603`](https://github.com/pohlm01/mtc/blob/main/mtc.go#L1603). Maybe it becomes more clear if I move this length prefix out of the TAI `MarshalBinary()` method?

Comment (bwesterb) [reply mtc.go:329]: Oh, you're right. Now my complain becomes the opposite: there are a few places where a second length prefix is included.

File: cmd/mtc/main.go [hunk 12]
```diff
@@ -601,38 +615,30 @@ func handleInspectCert(cc *cli.Context) error {
...
-					params.IssuerId,
-					anch.IssuerId(),
-				)
-			}
-			aa := c.Assertion.Abridge()
-			root, err := batch.ComputeRootFromAuthenticationPath(
-				proof.Index(),
-				path,
-				&aa,
+		if !reflect.DeepEqual(tai.Issuer, params.Issuer) {
```
Comment (bwesterb) [inline cmd/mtc/main.go:623]: Manual Equal function would be nice.

File: mtc.go [hunk 13]
```diff
@@ -488,12 +427,13 @@ func (p *CAParams) MarshalBinary() ([]byte, error) {
 	// TODO add struct to I-D
 	var b cryptobyte.Builder
 	b.AddUint8LengthPrefixed(func(b *cryptobyte.Builder) {
```
Comment (bwesterb) [inline mtc.go:429]: Eg here

Comment (pohlm01) [reply mtc.go:429]: To add even more to this confusion: Here it should be correct again. Please note the difference: In this case, I'm adding the raw bytes of the `RelativeOID` - just a type alias for `[]uint8` which does not (yet) implement a marshal method - while in the case discussed previously, I use the `MarshalBinary()` method of `TrustAnchorIdentifier`. Would you like me to align these usages, i.e., implement `MarshalBinary()` for `RelativeOID`?

File: cmd/mtc/main.go (hunk 12 above)
Comment (pohlm01) [reply cmd/mtc/main.go]: See [e5c15c7](https://github.com/bwesterb/mtc/pull/5/commits/e5c15c702786b1c74a86b82dff44ef4f6464beb8)

File: mtc.go (hunk 13 above)
Comment (bwesterb) [reply mtc.go:429]: Ah, ofc. Aligning them would be better.

**Review by bwesterb — APPROVED** [review]
Thanks! I'll merge when you're happy with final changes.

File: mtc.go (hunk 13 above)
Comment (pohlm01) [reply mtc.go:429]: See #6

**Pull request merged**